Add a notification mechanism for signaling when the mountpoint is ready

ariel-miculas opened the pull request:
Add the notify_pipe option which allows the user to specify the path to a named pipe. When the mountpoint is ready, a message will be written to this pipe: 's' for success and 'f' for failure. To avoid blocking the fuse process until the pipe is read, a new process is spawned which performs a blocking operation on the pipe. In case of mount failure, no additional process is created and the main process blocks until the pipe is read.

An example of operation is provided below:
```
set -x
FIFO=$(mktemp -u)
mkfifo "$FIFO"
./squashfuse_ll -o notify_pipe="$FIFO" -f /path/to/squashfs/archive /tmp/squash&
STATUS=$(head -c1 "$FIFO")
if [ "$STATUS" = "s" ]; then
	echo "Mountpoint contains:"
	ls /tmp/squash
else
	echo "Mounting squashfuse on /tmp/squash failed"
fi
```

Fixes #49

DrDaveD commented:
Is there any reason why this can't be done with squashfuse as well as squashfuse_ll?

ariel-miculas commented:
No, I'll add this to squashfuse also.

DrDaveD commented:
The code looks pretty good to me now (although the CI checks are running at the moment).  Please also add the new option to the `squashfuse.1` man page and add an automated test or tests.

vasi commented on ll.c:
```diff
@@ -432,6 +432,25 @@ void stfs_ll_op_statfs(fuse_req_t req, fuse_ino_t ino) {
+		return;
+	}
+
+	pid_t pid = fork();
```
This logic is a bit repetitive between the hl/ll cases. Maybe we can add a common function in fuseprivate?

vasi commented on ll.c:
```diff
@@ -432,6 +432,25 @@ void stfs_ll_op_statfs(fuse_req_t req, fuse_ino_t ino) {
+	sqfs_ll *ll = userdata;
+
+	if (!ll->fs.notify_pipe) {
+		return;
```
It's easy to miss that this will do the wrong thing if we ever add more logic to sqfs_ll_op_init. Maybe we can just make our common function (see below) return early if notify_pipe is null?

vasi commented on fuseprivate.c:
```diff
@@ -148,3 +150,34 @@ int sqfs_statfs(sqfs *sq, struct statvfs *st) {
 	return 0;
 }
+
+void notify_mount_ready(const char *notify_pipe, int status) {
```
Maybe instead of an int status, we can pass a constant char of 's' or 'f'? Feels a bit cleaner.

vasi commented:
Added some comments, but overall looks great!

ariel-miculas commented:
Thanks, @vasi. I've addressed your comments, updated the `squashfuse.1` manpage and modified some tests in `ll-smoke.sh.in` to use this new notification mechanism.

DrDaveD commented:
The new new smoke test unfortunately failed on the [freebsd check](https://api.cirrus-ci.com/v1/task/6241592026071040/logs/after.log).   Look for `FAIL:`.  It says
```
Image mounted successfully when it should have failed
```

ariel-miculas commented:
Thanks, I was looking for this log but couldn't find it. So it's probably fine to overmount squashfuse on top of an existing squashfuse mountpoint on freebsd, I need to change the test to make it fail the mount in other ways.

ariel-miculas commented:
I'm still awaiting approval for one workflow.

DrDaveD commented:
Thanks for the reminder, I'm sorry I didn't notice the push.  The good news was that freebsd passed.  I now approved the workflow.

DrDaveD commented:
This isn't fully working, see #112
